File: src/main/java/ru/sicampus/bootcamp2026/controller/UsersController.java
package ru.sicampus.bootcamp2026.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import ru.sicampus.bootcamp2026.dto.UserRegisterDTO;
import ru.sicampus.bootcamp2026.dto.UsersDTO;
import ru.sicampus.bootcamp2026.service.UsersService;

@RestController
@RequiredArgsConstructor
@RequestMapping("api/users")
public class UsersController {
    private final UsersService usersService;

    @GetMapping("/{id}")
    public ResponseEntity<UsersDTO> getUser(@PathVariable long id) {
        return ResponseEntity.ok(usersService.getUserById(id));
    }

    @PostMapping("/register")
    public ResponseEntity<UsersDTO> createUser(@RequestBody UserRegisterDTO dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(usersService.createUser(dto));
    }

    public ResponseEntity<UsersDTO> login(Authentication authentication) {
        return ResponseEntity.ok(usersService.getUserByUsername(authentication.getName()));
    }

    @PutMapping("/{id}")
    public ResponseEntity<UsersDTO> updateUser(@PathVariable long id, @RequestBody UsersDTO usersDTO) {
        return ResponseEntity.ok(usersService.updateUser(id, usersDTO));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteUser(@PathVariable long id) {
        usersService.deleteUser(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/username/{username}")
    public ResponseEntity<String> getUserByUsername(@PathVariable String username) {
        UsersDTO dto = usersService.getUserByUsername(username);
        return ResponseEntity.ok().body("User " + dto.getUsername() + " is registered");
    }

    @GetMapping("/paginated")
    public ResponseEntity<Page<UsersDTO>> getAllUsersPaginated(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        Pageable pageable = PageRequest.of(page, size);
        return ResponseEntity.ok(usersService.getAllUsersPaginated(pageable));
    }

    @GetMapping("/login")
    public ResponseEntity<String> checkAuth() {
        return ResponseEntity.ok("Auth completed");
    }
}

File: server/src/main/kotlin/com/nikshet/industrialbackend/contoller/UsersController.kt
package com.nikshet.industrialbackend.contoller

import com.nikshet.industrialbackend.domain.TokensEntity
import com.nikshet.industrialbackend.dto.request.UpdateProfileRequest
import com.nikshet.industrialbackend.dto.request.UserRegistrationRequest
import com.nikshet.industrialbackend.dto.response.UserInfoResponse
import com.nikshet.industrialbackend.providers.JwtTokensProvider
import com.nikshet.industrialbackend.service.UsersService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

@RestController
@RequestMapping("/users")
class UsersController(
    private val usersService: UsersService,
    private val tokensProvider: JwtTokensProvider
) {
    @PostMapping("/registration")
    fun registerUser(@Valid @RequestBody request: UserRegistrationRequest): ResponseEntity<TokensEntity> {
        val user = usersService.registerUser(
            phoneNumber = request.phoneNumber,
            fullName = request.fullName,
            department = request.department,
            password = request.password
        )

        val payload: Map<Any, Any> = mapOf(
            "sub" to user.id.toString(),
            "phone" to user.phoneNumber
        )

        val tokens = tokensProvider.generateTokensPairs(payload)
        return ResponseEntity.status(HttpStatus.CREATED).body(tokens)
    }

    @PutMapping("/profile")
    fun updateProfile(
        @AuthenticationPrincipal userId: UUID,
        @Valid @RequestBody request: UpdateProfileRequest
    ): ResponseEntity<UserInfoResponse> {
        val updatedUser = usersService.updateUserProfile(userId, request)
        return ResponseEntity.ok(UserInfoResponse.from(updatedUser))
    }

    @GetMapping("/me")
    fun getProfile(@AuthenticationPrincipal userId: UUID): ResponseEntity<UserInfoResponse> {
        val user = usersService.getUserById(userId)

        return ResponseEntity.ok(UserInfoResponse.from(user))
    }

    @GetMapping
    fun getUsers(): ResponseEntity<List<UserInfoResponse>> {
        val users = usersService.getAllUsers()

        return ResponseEntity.ok(users.map { UserInfoResponse.from(it) })
    }
}